### PR TITLE
Gather per-example tensors before reducing logged means

### LINKED ISSUE #7010: Metrics are reduced to a rank-local scalar before gather_for_metrics, biasing the final batch
### Summary

Several metrics are reduced to a rank-local scalar before being handed to
`Accelerator.gather_for_metrics`. That defeats the de-duplication `gather_for_metrics` exists to do,
so on the final batch of an epoch the reported value is biased whenever the dataset does not divide
evenly across ranks.

`trl/trainer/reward_trainer.py` is affected, so this is not confined to `trl/experimental`.

### Why the pre-reduction breaks it

`gather_for_metrics` de-duplicates the padding accelerate added to the last batch by slicing the
gathered tensor along the batch dimension. From the installed accelerate source:

```python
elif self.gradient_state.remainder > 0:
    return tensor[: self.gradient_state.remainder]
```

That only works on a per-example tensor. A rank-local scalar has no batch dimension left to slice,
and the duplicated example is already folded into the number, so the truncation cannot remove it.

### The size of the error

Nine real examples, two ranks, local batch size five, so accelerate pads the second rank by one.
Exact rationals:

| what the padding repeats | reported | true | error |
| --- | --- | --- | --- |
| the first example | 18/5 = 3.6 | 4 | -0.4 |
| the last example | 22/5 = 4.4 | 4 | +0.4 |

A ten percent error on that batch, with the sign set by which element accelerate repeats. As a
control, with equal counts and no padding the mean of the rank means equals the true mean exactly, so
the bias comes specifically from the padded batch.

### The same file does it both ways

`trl/experimental/cpo/cpo_trainer.py`, four lines apart in one function:

```python
# correct: the per-example tensor reaches the de-duplicator, the mean comes after
self.accelerator.gather_for_metrics(policy_chosen_logps).detach().mean().item()

# biased: the mean happens first, so there is nothing left to de-duplicate
self.accelerator.gather_for_metrics(policy_rejected_logits.detach().mean()).mean().item()
```

That is what makes this look like an oversight rather than a deliberate convention.

### Sites

Found with an AST scan for a `.mean()` reaching `gather_for_metrics` or `gather_mean`, either inline
or through a variable assigned one line earlier. 21 sites in 7 files:

| file | sites |
| --- | --- |
| `trl/trainer/reward_trainer.py` | 775, 779 |
| `trl/experimental/xpo/xpo_trainer.py` | 387, 388, 394, 395, 411, 415 |
| `trl/experimental/online_dpo/online_dpo_trainer.py` | 1264, 1266, 1275, 1279 |
| `trl/experimental/ppo/ppo_trainer.py` | 905, 906, 908, 911 |
| `trl/experimental/cpo/cpo_trainer.py` | 906, 909 |
| `trl/experimental/orpo/orpo_trainer.py` | 864, 867 |
| `trl/experimental/tpo/tpo_trainer.py` | 708 |

`reward_trainer.py:774` is the clearest case: `(rewards_chosen > rewards_rejected).float().mean()`
reduces one boolean per pair to a scalar, then gathers it, so `accuracy` carries the duplicate.

I have not checked all 21 against their surrounding code. Anywhere the quantity is already global
rather than per-example, the ordering does not matter and the site can be dismissed.

### Suggested fix

Pass the per-example tensor and reduce after the gather, which is what `cpo_trainer.py:902` already
does:

```python
self.stats["logps/chosen"].append(gather_mean(chosen_model_logprobs))
```

### How I found it

A cross-model reviewer raised two of these sites on an unrelated PR. I confirmed the mechanism
against the installed accelerate source, computed the bias exactly, and swept the tree for the rest.
Happy to open the PR if the direction looks right.

### PR DESCRIPTION
**Pull request by behroozazarkhalili**
Several trainers reduced metrics before gathering them, preventing removal of duplicated examples in the final distributed batch. I changed the affected paths to gather per-example values before reduction and corrected related padding and label handling. Tests that simulate duplicate removal pass and fail when the changes are reverted, but I did not run multiple processes.
Fixes #7010.

---

> [!NOTE]
> **Medium Risk**
> Touches metric logging across many RLHF trainers; training losses are mostly unchanged but logged dashboards and early-stopping on metrics could shift on uneven distributed datasets.
> 
> **Overview**
> Fixes biased training/eval metrics when dataset size does not divide evenly across distributed ranks. **`gather_for_metrics`** only strips Accelerate’s duplicated last-batch examples along the batch dimension, so metrics that were **averaged (or summed) before the gather** kept the extra padding example baked in.
> 
> Across **DPO, KTO, Reward, CPO, ORPO, TPO, XPO, Online DPO** (and related paths), logging now passes **one value per example** (or per prompt where batches stack chosen/rejected/reference or two completions), gathers, then reduces. Entropy and average logits use **per-row token sums and counts**; stacked batches are **folded** so one row maps to one example before gather. **CPO/ORPO** expose **per-example NLL sum and token count** from forward; the training loss is unchanged but **`nll_loss`** is computed after gather. **Online DPO** aggregates KL, entropy, scores, and EOS flags **per prompt** and masks padding in KL/entropy. **KTO** casts labels to **bool**, gathers KL per row, and **scatters** chosen/rejected rewards/logps into full-batch tensors so single-class batches and **bfloat16** dtypes work with gather. **Reward** and **XPO** log accuracy, margins, and rewards from per-pair tensors instead of pre-reduced scalars.
> 
> **Tests** add **`drop_last_for_metrics`** in `testing_utils` (mimics Accelerate de-duplication) and oracle tests that assert logged values match the mean over non-outlier examples when the last batch row is an outlier.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 000fc99f7189986968c3aee263a9eff684ff9144. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/trainer/dpo_trainer.py [hunk 1]
```diff
@@ -1607,10 +1607,14 @@ def _compute_loss(self, model, inputs, return_outputs):
...
-        total_tokens = mask.sum()
+        entropy_sum = (per_token_entropy * mask).sum(dim=1)
+        total_tokens = mask.sum(dim=1)
 
-        # Gather counts across ranks and weight-average
+        # `gather_for_metrics` de-duplicates examples, but the batch stacks chosen and rejected rows.
+        entropy_sum = entropy_sum.view(2, -1).sum(0)
+        total_tokens = total_tokens.view(2, -1).sum(0)
+
+        # Gather per-example sums and counts across ranks, then weight-average
```
[the remark this answers is not in the document]
Comment (behroozazarkhalili) [reply trl/trainer/dpo_trainer.py:1617]: Confirmed. TPO copies DPO's metric block and got the per-example shape in the earlier round while DPO and KTO did not, and DPO's reward and log-prob sites used plain `gather`, which never removes the padding at all. Fixed in c80da9db: DPO gathers per-example logit sums, token counts, correct-token counts, rewards, accuracies, margins and log-probs in both the standard and the Liger block; KTO gathers the KL, the token count, the chosen and rejected counts as one boolean per row, the logits and the rewards and log-probs scattered into a row-aligned tensor. The DPO and KTO oracles now cover every one of those metrics and the DPO fixture no longer stubs `gather`; on effd19be the DPO oracle fails on that stub and both KTO cases keep the duplicated KL. Two mutants that re-collapse the DPO logits and the token accuracy to scalars fail the oracle (6.67 vs 5.0 and 0.667 vs 1.0). The full DPO and KTO suites pass on a compute node (199 passed, 5 skipped). One limit stays: the Liger logits in DPO and the Liger logits, log-probs and rewards in KTO come back from liger_kernel as 0-d sums, so those sites keep the scalar until Liger returns row-aligned values; the body lists it.

File: trl/trainer/kto_trainer.py [hunk 2]
```diff
@@ -1647,20 +1645,25 @@ def _compute_loss(self, model, inputs, return_outputs):
         all_num_chosen = self.accelerator.gather_for_metrics(num_chosen).sum().item()
         all_num_rejected = self.accelerator.gather_for_metrics(num_rejected).sum().item()
 
+        chosen_rewards_sum = torch.zeros_like(completion_logps).index_copy(0, chosen_idx, chosen_rewards)
+        rejected_rewards_sum = torch.zeros_like(completion_logps).index_copy(0, rejected_idx, rejected_rewards)
```
[the remark this answers is not in the document]
Comment (behroozazarkhalili) [reply trl/trainer/kto_trainer.py:1653]: Confirmed by execution: `torch.zeros(3, dtype=torch.bfloat16).index_copy(0, empty_idx, torch.Tensor([]))` raises `index_copy_(): self and source expected to have the same dtype`, and the KTO loss returns exactly that float32 default when a batch has no chosen or no rejected rows. Fixed in 3da7a566: both reward tensors are cast to the row tensor's dtype before the scatter. New test `test_metrics_with_a_single_label_class_in_bfloat16` runs a bfloat16 batch whose rows are all chosen, then all rejected; on f5e8e9a9 both cases raise the dtype error, on 3da7a566 they pass and the present class's reward matches the mean.

File: trl/trainer/kto_trainer.py [hunk 3]
```diff
@@ -1495,8 +1495,8 @@ def _compute_loss(self, model, inputs, return_outputs):
         batch = {k: (v.to(self.accelerator.device) if isinstance(v, torch.Tensor) else v) for k, v in inputs.items()}
 
         labels = batch["label"]
-        num_chosen = labels.sum().to(self.accelerator.device)
-        num_rejected = (len(labels) - num_chosen).to(self.accelerator.device)
+        num_chosen = labels.to(self.accelerator.device)
+        num_rejected = ~num_chosen
```
[the remark this answers is not in the document]
Comment (behroozazarkhalili) [reply trl/trainer/kto_trainer.py]: Confirmed: `~torch.tensor([1, 0])` gives `[-2, -1]`, and the metric path read `batch["label"]` raw while the loss path already casts it to bool. Fixed in 0b4d8dbc: the metric path casts to bool before the inversion. The oracle now runs over bool and int64 labels; on 4ee8527d the int64 rows fail with an IndexError from the negative mask, and all rows pass now.

**Comment by behroozazarkhalili** [comment]
Closing this one myself. It came from reading the code, not from a failure in a real training run, and per the discussion on #6862 that isn't the bar for this repo.

@kashif, thanks for the commits you put into this. If you think it's worth keeping, say the word and I'll reopen it.
